utils/utils.py

```diff
@@ -0,0 +1,387 @@
+import json
+import logging
+from dataclasses import asdict, dataclass
+from datetime import datetime, timezone
+from http import HTTPStatus
+from typing import Dict, List
+
+import boto3
+import requests
+from jose import jwk, jwt
+from jose.utils import base64url_decode
+from requests_oauth2client import OAuth2Client
+from retry import retry
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+
+@dataclass
+class Identifier:
+    entityId: str
+    entityType: str
+
+@retry(tries=3, delay=2)
+def identity_login(identity_url: str, username: str, password: str) -> str:
+    try:
+        print('identity url:', identity_url)
+        oauth2client = OAuth2Client(
+            token_endpoint=f'{identity_url}/oauth2/platformtoken',
+            auth=(username, password),
+            timeout=10)
+        token = oauth2client.client_credentials(scope="", resource="")
+        return str(token)
+    except (Exception) as ex:
+        print(ex.error)
+        if "access_denied" in ex.error:
+            raise Exception("Access Denied")
+
+
+def _get_data_entities(token_claims: Dict,
+                       user_attributes: Dict = None) -> List:
+    """ Get data entities from token claims and user attributes
+
+    Parameters:
+        token_claims:
+        user_attributes:
+
+    Returns:
+
+    """
+    data_entities: List[Dict] = []
+    if not token_claims:
+        return data_entities
+
+    # add roles from token
+    for role in token_claims['user_roles']:
+        data_entities.append({
+            'identifier':
+            asdict(Identifier(entityType='UserGroup', entityId=role))
+        })
+
+    # add user and role parents
+    user_entity = {
+        'identifier':
+        asdict(Identifier(entityType='User', entityId=token_claims['sub'])),
+        'parents': []
+    }
+    if user_attributes:
+        user_attributes_dict = {}
+        for attribute in user_attributes:
+            user_attributes_dict[attribute] = {
+                'string': user_attributes[attribute]
+            }
+        user_entity['attributes'] = user_attributes_dict
+
+    for role in token_claims['user_roles']:
+        user_entity['parents'].append(
```

Comment:
consider:
user_entity['parents'] = [asdict(Identifier(entityType='UserGroup', entityId=role))) for role in token_claims['user_roles']]

data_entities.append(user_entity)

Reply:
updated